쥬스 메이커 [STEP 2] marisol, mmim

JoSH0318 opened the pull request:
@stevenkim18 안녕하세요🥳 스티븐! STEP2 PR입니다. 잘부탁드립니다.

## 📌UML
<img width="6112" alt="JuiceMaker STEP2_UML " src="https://user-images.githubusercontent.com/88810018/156318394-1e12d13e-cb75-46b0-9742-efa30739cbc0.png">

## 📌주요기능
### class OrderViewController
- `func showCurrentStock()`: 현재 과일 재고를 `stockLabels`에 보여주는 기능
- `func presentStockViewController()`: `StockViewController`로 화면 전환해주는 기능
- `func orderButtonsClicked()`: IBAction Collection으로 구현, 버튼별 동작을 매칭해주는 기능 + 재고 갱신하기 위한 `showCurrentStock()` 호출
- `func alertOrderCompletion()`: 주문 완료시 생성되는 알럿 기능
- `func alertOutOfStock()`: 재고 부족 시 생성되는 알럿 기능

## 📌고민한점과 해결방법 / 이렇게 해결한 이유
### 과일 재고를 나타내는 Label을 Outlet Collection으로 사용

- 의미가 비슷하고 단순 나열되어 있는 5개의 과일 Label을 하나씩 만들어주지 않고, Outlet Collection으로 만들었습니다. 코드를 간결하게 할 수 있다는 장점이 있어서 Outlet Collection 기능을 사용했습니다.
    
### 화면 전환을 modal로 선택
- 쥬스를 주문하는 화면에서 재고를 추가하는 화면이 쭉 이어지는 흐름이 아니라고 생각했습니다. 마치 edit과 같은 성격의 화면전환이라 생각했고 modal을 사용하는 것이 적절하다 판단했습니다.
    
### 여러 IBAction을 하나로 묶어서 사용
- 쥬스 버튼마다 IBAction를 생성할 경우, 비슷한 기능을 하는 IBAction을 7가지를 생성해야 해서 IBAction을 하나로 묶어 하나의 메서드에서 sender를 구분해서 쥬스를 만들도록 했습니다. UIButton마다 tag를 사용해서 구분하였고, switch로 sender.tag를 돌면서 쥬스를 만들고 재고를 업데이트 하도록 했습니다.

### static을 이용한 singleton pattern 채택
- 과일재고라는 단하나의 자원을 여러 객체에서 공유하는 컨셉이라고 생각했고, 이에 singleton이 적절하다 판단했습니다. 마침 학습활동으로 singleton에 대해 학습했고, 이번 기회에 적용해 사용해보는게 좋을 것 같아 사용했습니다. 이전에 singleton과 static 사용이 미숙하여 좀 더 학습하고 보안해보았습니다. 
   
### enum OrderError의 사용
- 이번 프로젝트를 보면 error의 종류는 '재고부족', '알수없는 error' 두 가지 밖에 없습니다. 때문에 굳이 error를 enum으로 만들어주어야만 하나...라는 고민을 했습니다. 하지만 주문 버튼을 누르고 주문 과정에서 재고가 있다면 정상 동작해야하고, 재고가 부족하다면 주문 과정을 멈추고 오류를 핸들링해줘야만 했습니다. 이러한 이유로 `enum OrderError: Error {}`가 있어야 우리가 원하는 do-catch로 로직을 구현할 수 있을 거라 판단했습니다. 더 나은 방법, 또는 부족한 점이 있다면 알려주시면 감사하겠습니다.🥲

### enum Alert의 사용
- Alert을 사용하게 되면서 title과 message를 매직 리터럴을 사용하게 되었고, 만약 프로젝트 규모가 크다면 Alert의 내용을 변경할 때 하나하나 바꿔줘야한다는 생각이 들었습니다. 이 두가지 문제점을 해결하기 위해 enum Alert을 사용하게 됐습니다. 

## ❓궁금한점❓
- 쥬스 주문하는 화면에서 버튼들을 모두 묶어 하나의 메서드에서 처리하도록 하고, tag로 구분해주었습니다. tag라는 기능을 처음 사용하는 거라, 혹시 부족한 점이 있다면 알려주시면 감사하겠습니다.🙇‍♀️
- 일반적으로 IBOutlet Collection을 사용하는 지 궁금합니다. 저희가 판단하기로 비슷한 성격으로 나열된 Label이나 Button을 Collection으로 묶는다면 코드가 간결해진다고 판단했지만, 현직에서 자주 사용되는지, 또는 가독성이 떨어지는 문제가 있는 것은 아닌지 궁금합니다!

marisol-develop committed:
[feat]: IBOutlet, IBAction 연결

과일 라벨과 쥬스만들기 버튼을 viewController에 연결

JoSH0318 committed:
[feat]: add showCurrentStock(), combine outlets

showCurrentStock() method: 현재 재고를 형변환하여 라벨에 저장하는 메서드

marisol-develop committed:
[feat]: juiceOrderButtonClicked 기능 구현

버튼 클릭된 쥬스별로 쥬스 제조하고 현재 재고를 보여주는 기능 구현

JoSH0318 committed:
[feat]: add alert methods, move order() method to viewController

- 주문완료, 재고부족 알럿 메서드 구현
- order() 메서드 viewController로 이동
- enum Juice에 extension 생성(한글명)
- enum OrderError 의 extension 삭제

marisol-develop committed:
[feat]: connect MainVC and StockVC, add viewWillAppear

- 세그로 MainVC와 StockVC 연결
- viewWillAppear 오버라이딩

JoSH0318 committed:
[feat]: delete second navigationController, add cancelButton method

- 두번째 내비게이션 컨트롤러 삭제 -> 뒤로가기 버튼, 타이틀 라벨 커스텀
- 뒤로가기 버튼 기능 추가

marisol-develop committed:
[refactor]: combine juiceOrderButtons to collection

쥬스별 주문 버튼을 IBActionCollection으로 통합

JoSH0318 committed:
[feat]: add method, Enum

- transformTypeToString() : dictionary의 Int 값을 String?로 바꿔주는 메서드
- enum Alert : 알럿 관련된 타이틀, 메시지를 위한 enum
- 이에 따른 부분 로직 수정

marisol-develop committed:
[refactor]: modify view transition

- segue -> present로 수정

JoSH0318 committed:
[refactor]: rename, delete file / method / white space

marisol-develop committed:
[del]: delete viewWillAppear()

- viewWillAppear 메서드 삭제
- transformTypeToString -> transformIntToString로 수정

JoSH0318 committed:
[feat]: add private, final keyword

marisol-develop committed:
[rename]: modify MainVC name to OrderViewController

- MainViewController를 OrderViewController로 수정

JoSH0318 committed:
[docs]: update README.md

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+
+import UIKit
+
+final class OrderViewController: UIViewController {
```
뷰컨에 이름을 주신 것은 정말 잘하셨어요!!! 

개인적으로 저라면 main 뷰컨이라고 할 것 같아요 ㅎㅎ

JoSH0318 replied:
해당 view의 역할은 `주문`을 하는 것이라고 생각했습니다! 때문에 view controller의 이름 또한 view의 역할을 따라 OrderViewController로 짓게 됐습니다 ☺️

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+//
+//  JuiceMaker - ViewController.swift
```
이런 디테일도 신경써주시면 좋아요 ㅎㅎ 

파일 이름이랑 주석 이름이랑 동일하게 수정해주세요! ㅎㅎ

marisol-develop replied:
파일이름과 주석이름 동일하게 수정했습니다!😀

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+//
+//  JuiceMaker - ViewController.swift
+//  Created by yagom. 
```
다른 파일에는 밈 마리솔 이름이 들어가 있는데 여기는 야곰이네요 ㅋㅋ

marisol-develop replied:
이것도 밈 마리솔로 수정했어요 ㅎㅎ

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+    private let juiceMaker = JuiceMaker()
+    
+    @IBOutlet var stockLabels: [UILabel]!
+    @IBOutlet var juiceOrderButtons: [UIButton]!
```
1. 버튼과 라벨을 각각 하나씩 변수로 두었을 때랑 배열로 관리했을 때 장단점이 무엇이 있을까요??
2. 왜 @IBOutlet 은 자동적으로 ?가 아닌 !로 선언 될까요????

marisol-develop replied:
1. 각각 하나씩 변수로 두었을 때는 weak으로 변수를 선언할 수 있어 순환참조를 방지할 수 있다는 장점이 있고, 다만 하나하나 연결해줘야 하기 때문에 코드가 조금 길어질 수 있을 것 같습니다. 배열로 관리했을 때는 코드가 짧아지지만, 인덱스로 접근해야하기 때문에 가독성이떨어진다는 단점이 있을 것 같네요!
2. IBOutlet은 스토리보드의 뷰 객체와 Interface Builder를 연결하는데, 뷰 객체가 생성되고 난 뒤로부터는 항상 값이 들어있을 것이기 때문에 자동으로 !로 선언되는 거라고 생각합니다!

질문에 대해 생각해보았는데 혹시 잘못되었다면 말씀 부탁드립니다.😃

stevenkim18 replied:
와 순환참조까지 알아보셨다니 대단합니다!!! 나중에 배울 거니 지금은 깊에 안보셔도 됩니다!!

그리고 2번에 대한 답변도 저의 생각과 유사해 놀랐습니다.

IBOutlet과 같은 경우에는 애플에서 개발자가 반드시 뷰를 잘 연결해 줘야만 빌드가 되기 때문에 !로 선언한 것 같습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+    
+    @IBOutlet var stockLabels: [UILabel]!
+    @IBOutlet var juiceOrderButtons: [UIButton]!
+    @IBOutlet weak var viewTransitionButton: UIBarButtonItem!
```
이 변수 이름만 봤을 때 어떤 뷰로 전환 되는 버튼 인지 잘 모르겠어요!

어떻게 이름을 바뀌주면 좋을까요?

marisol-develop replied:
재고관리하는 뷰로 이동한다는 의미로, moveToStockViewButton으로 수정했습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+                stockLabels[3].text = transformIntToString(.kiwi)
+            default:
+                stockLabels[4].text = transformIntToString(.mango)
+            }
```
코드만으로 0번째 text가 딸기 라벨인지 어떻게 알 수 있을까요?

그리고 default면 무조건 망고 라벨일까요??

marisol-develop replied:
stockLabels를 컬렉션으로 묶어 사용해서 인덱스로 접근했어야 했는데, 가독성이 좋지 않은 것 같아서 컬렉션을 풀어서 라벨을 각각 지정해주었습니다. 그리고 마지막 default를 case .mango로 변경했습니다.😊

stevenkim18 replied:
분명히 명시해주시는게 미래의 나를 위해서 그리고 유지보수시 애러 방지를 위해서 더 좋습니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+            return nil
+        }
+        return String(currentStock)
+    }
```
함수의 이름이 int to String인데 매개변수로 `Fruit`이 들어가는 것이 이상해요!

그리고 transform이라는 단어를 사용하셨는데 사전적 의미는 무엇인가요?

변형, 변환하다에는 change, convert등 다양한 단어가 있는데 transform를 사용하신 이유가 있을까요?

marisol-develop replied:
Fruit 타입을 매개변수로 받는 것이 어색해서, 네이밍을 convertFruitStockToString로 수정했습니다.

stevenkim18 replied:
함수 이름을 지으실때도 영어 단어의 의미를 한번 찾아보시는 것도 도움이 됩니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+        self.presentStockViewController()
+    }
+    
+    private func presentStockViewController() {
+        guard let stockViewController = self.storyboard?.instantiateViewController(withIdentifier: "stockViewController") as? StockViewController else {
+            return
+        }
+        stockViewController.modalTransitionStyle = .coverVertical
+        stockViewController.modalPresentationStyle = .automatic
+        self.present(stockViewController, animated: true, completion: nil)
+    }
```
지금 코드로 구현하신 것을 스토리보드 만으로도 구현할 수 있습니다!! 어떻게 하면 될까요?

JoSH0318 replied:
스토리보드에서 직접 다음화면에 `drag&drop`하고 `Present Modality `를 선택하는 방법이 있습니다!

저희 코드에서 `presentStockViewController() `함수는 재고수정화면으로 넘어가는 버튼 `moveToStockViewButtonClicked`과 알럿을 통해 재고수정화면으로 넘어가는 `alertOutOfStock()`에서 호출되고 있습니다. 
만약 스토리보드에서 `drag&drop`, `Present Modality `를 하게 되면 버튼을 통한 화면전환을 할 수 있습니다. 하지만 Alert을 통한 화면 전환은 코드로 구현해주어야만 했습니다. 따라서 공통된 화면전환 코드를 만들고 각 메서드에 호출하는 방법을 선택했습니다.

stevenkim18 replied:
앗 네네!! 말씀하신 drag&drop 이 Segue를 사용하는 방법입니다!!

다양한 방법이 있다는 것을 말씀드리고 싶었습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/OrderViewController.swift:
```diff
@@ -0,0 +1,111 @@
+            order(.mangoJuice)
+            showCurrentStock()
+        }
+    }
```
이 부분도 1이라는 것이 어떻게 딸기 바나나 주스와 연결이 될까요?

defalut면 무조건 망고주스를 만들어야 하나요?

marisol-develop replied:
tag를 쓰면 가독성이 아무래도 많이 떨어지는 문제가 있어서 UIButton을 각각 따로 설정해주고 case를 button으로 구분해주었습니다. 그리고 showCurrentStock이 중복되는 코드라서, switch문 밖에 한번만 호출해주었습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Alert.swift:
```diff
@@ -0,0 +1,52 @@
+            return "나왔습니다! 맛있게 드세요!"
+        }
+    }
+}
```
문자열 상수를 enum으로 관리해주신 것은 정말 잘하셨습니다!!!!

JoSH0318 replied:
👍

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,46 +6,12 @@
 import Foundation
 
-class FruitStore {
+final class FruitStore {
```
오 final 키워드를 사용하셨네요!!!

사용했을 때와 아닐 때 차이는 무엇인가요?

marisol-develop replied:
final 키워드는 더 이상 해당 클래스를 상속 하거나, 상속을 통해 재정의될 필요가 없다고 판단한 경우 class 앞에 붙여 상속을 방지할 수 있습니다. 저희의 경우 OrderViewController와 StockViewController가 더이상 상속이나 상속에 따른 재정의가 필요하지 않다고 생각하여 final 키워드를 붙였습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,46 +6,12 @@
 
-class FruitStore {
+final class FruitStore {
+    static let shared = FruitStore()
```
주스메이커의 프로퍼티로 한번만 참조를 하셨는데 싱글톤으로 구현하신 이유가 궁금합니다!

marisol-develop replied:
step2까지는 한번만 참조를 하지만, 추후에 StockViewController에서도 과일의 현재 재고를 파악하기 위해 참조할 필요가 있다고 생각해서 싱글턴으로 구현했습니다.

stevenkim18 replied:
토요일 새벽에 잠시 구두로 말한 것 처럼 딜리게이트 패턴을 적용해보시면서 싱글톤을 사용하지 않는 방법도 고민해보시면 좋을 것 같습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,46 +6,12 @@
-    init() {
+    private init() {
         let defaultQuantity = 10
         Fruit.allCases.forEach{ inventory[$0] = defaultQuantity }
```
forEach를 사용하셨네요!! 잘하셨습니다!! 

for - in 어떤 차이가 있나요??

JoSH0318 replied:
1. for문은 조건 반복문 / foreach 고차함수
   - for 문은 조건을 직접 구현해서 조건에 따라 반복합니다. 따라서 조건에 따라 continue, break을 사용할 수 있습니다.
   - 하지만 foreach는 함수이기 때문에 파라미터로 클로저를 작성해서 넘겨줍니다.
   - 또한 foreach는 반복문이 아니기 때문에 continue, break를 사용하지 못합니다. 
   - for문은 반복할 횟수, 범위를 정해줄 수 있습니다. 하지만 foreach는 배열의 element를 모두 접근해야 합니다. 한 마디로 for문의 반복횟수를 제어할 수 있고, foreach는 불가능합니다.(사실 아직까진 foreach의 횟수를 제어하는 방법을 못찾은 것 같기도 합니다 😂)
   
2. return의 차이
   - for문에선 바디에서 반복하는 중 return을 만나면 for문을 종료하게 됩니다. (함수안에 있다면 함수 자체가 종료)
   - foreach는 return을 만나도 클로저가 종료될 뿐 foreach문 자체가 종료되지 않습니다. 따라서 element만큼 계속하여 반복하게 됩니다.

제가 경험하고 찾은 차이는 여기까지 입니다 😂

stevenkim18 replied:
공부해보시라도 드린 질문이었습니다!! ㅎㅎ

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -34,4 +36,23 @@ enum Juice {
+        case .mangoKiwiJuice:
+            return "망키 쥬스"
+        }
+    }
```
`Juice`타입에서 `recipe`와 `name`을 extension으로 분리하신 이유가 있을까요??

JoSH0318 replied:
부연 설명의 느낌으로 가독성을 위해서 extension을 사용하여 recipe와 name을 구분했습니다. ☺️
name 같은 경우는 CustomStringConvertible을 이용하여 description으로 수정했습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -34,4 +36,23 @@ enum Juice {
         }
     }
+    
+    var name: String {
+        switch self {
+        case .strawberryJuice:
+            return "딸기 쥬스"
+        case .bananaJuice:
+            return "바나나 쥬스"
+        case .kiwiJuice:
+            return "키위 쥬스"
+        case .pineappleJuice:
+            return "파인애플 쥬스"
+        case .strawberryBananaJuice:
+            return "딸바 쥬스"
+        case .mangoJuice:
+            return "망고 쥬스"
+        case .mangoKiwiJuice:
+            return "망키 쥬스"
+        }
+    }
```
이렇게 네임으로 사용하려면 CustomStringConvertible를 공부해보셔도 좋을 것 같아요!

marisol-develop replied:
말씀해주신대로 Juice 열거형에 CustomStringConvertible 프로토콜을 채택하여 description 프로퍼티을 정의해서 사용해보았습니다!

stevenkim18 replied:
사용할 때 코드가 간결해지는 것을 확인하셨나요?

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -2,7 +2,7 @@
+//  Created by marisol, mmim on 2022/02/22.
 //
 
 import Foundation
```
`Fruit`에 `CaseIterable`를 사용하신 것은 정말 잘하셨습니다!

(해당 코멘트가 선택이 되지 않아서 바로 위부분을 태그 했습니다)

stevenkim18 reviewed (commented):
안녕하세요! 밈 마리솔 스탭2까지 오시느라 고생많으셨습니다.

제가 최대한 24시간 내에 해드리고 싶은데 ㅜㅜ 그러지 못해서 죄송해요.

아 하나 부탁드리고 싶은 것은 다음 스탭에서는 메인 브런치에 머지해서 PR 날려주시면 감사할게요!! 클론 받기가 그게 편해서 그렇습니다!! 

### UML에 대해서

- UML을 너무 잘 그려주셨습니다!! 그리시는데 시간이 얼마나 걸리셨는지 궁금하네요 ㅎㅎ 시간이 많이 걸린다면 굳이 예쁘게 하실 필요는 없고 그냥 종이에 연필로 어떤 객체가 있고 어떤 관계를 가지는지만 간단하게 그려주시면 효율성이 높을 것 같아요!
- 그리고 view에 대해서는 UML 보다 구현하신 화면의 사진이나 gif를 첨부하시면 제가 더 쉽게 인지할 수 있을 것 같습니다!!
- 어떻게 하면 여러분들이 고생하시면서 한 결과물을 잘 보여줄 수 있을 지 끊임 없이 다양하게 고민해주세요!

### ****과일 재고를 나타내는 Label을 Outlet Collection으로 사용****

- 장단점이 확실히 있습니다! 알고 넘어가시면 좋아요!

### ****화면 전환을 modal로 선택****

- 스토리보드를 보면서 가장 질문하고 싶었던 것이었는데 원레 `재고 추가` 뷰컨이 네비게이션 컨트롤러로 감쌓아져있지 않았나요?? 수정하신 이유가 궁금합니다!

### ****여러 IBAction을 하나로 묶어서 사용****

- 밑에 코드 리뷰에서도 적었지만 tag를 사용하면 장단점이 확실 있습니다.
- 개인적으로는 tag를 숫자로 사용해서 단점이 더 많이 부각되는 것 같습니다. 어떤 숫자가 어느 버튼에 해당하는지 tag으로 판단하기는 어려운 것 같습니다.

### ****static을 이용한 singleton pattern 채택****

- 이 부분도 밑에 코드에서 언급했지만 `juiceMaker` 에서 한번 프로퍼티로 선언만 하고 다른 곳에는 전혀 쓰이지 않았는데 굳이 싱글톤으로 선언하신 이유가 궁금합니다!

### ****enum OrderError의 사용****

- 에러 케이스가 두개라고 해도 해당 에러에 대해서 확실한 분기 처리나 에러 메세지가 명시되어야 한다면 애러 타입을 사용하셔도 무방하다고 생각합니다.
- 그런것이 아니면 굳이 if else로도 처리 가능하다고 생각합니다.
- 지금 주신 코드에서도 `outOfStock` 일때만 처리를 해주셔서 크게 Error 타입을 사용하신 이유에 대해서 뚜렷하게 보이지는 않는 것 같습니다.

### ****enum Alert의 사용****

- 너무 잘하셨습니다! 조금 더 경험이 쌓이면 한개의 함수로도 처리하실 수 있을실 것 같아요!

> 쥬스 주문하는 화면에서 버튼들을 모두 묶어 하나의 메서드에서 처리하도록 하고, tag로 구분해주었습니다. tag라는 기능을 처음 사용하는 거라, 혹시 부족한 점이 있다면 알려주시면 감사하겠습니다.
> 
- 만약 제가 tag를 사용할 것이라면 enum에 rawValue로 해당 tag 값들을 줄 것 같습니다. 그러면 명시적으로 tag(int)를 사용하고도 코드에 어떤 주스의 버튼인지를 알게 할 수 있습니다.

> 일반적으로 IBOutlet Collection을 사용하는 지 궁금합니다. 저희가 판단하기로 비슷한 성격으로 나열된 Label이나 Button을 Collection으로 묶는다면 코드가 간결해진다고 판단했지만, 현직에서 자주 사용되는지, 또는 가독성이 떨어지는 문제가 있는 것은 아닌지 궁금합니다!
> 
- 때에 따라서 다르겠지만 저는 많이 사용해보지는 않았습니다. 기획에서 정말 단순히 모든 버튼이 거의 비슷한 기능을 하는 것은 제가 경험한 서비스에서는 잘 보지 못했습니다.
- 하지만 제가 경험이 부족하기에 많이 사용할 수도 있습니다.

marisol-develop committed:
[refactor]: change outlet collection to single outlet

- UILabelCollection이었던 stockLabels을 개별 UILabel로 수정
- UIButtonCollection이었던 juiceOrderButtons를 개별 UIButton으로 수정
- 이에 따른 tag 제거 및 collection index 사용 수정

JoSH0318 committed:
[refactor]: add enum Alert case, extension OrderViewController

- enum Alert에 unknownError 추가
- extension OrderViewController해서 Alert 기능 메서드를 몰아넣음
- Created by / Copyright 주석 수정

marisol-develop committed:
[refactor]: use CustomStringConvertible, change method name

- Juice의 extension에서 CustomStringConvertible 사용
- transformIntToString -> convertFruitStockToString으로 네이밍 변경
- viewTransitionButtonClicked -> moveToStockViewButtonClicked으로 네이밍 변경

JoSH0318 committed:
[improve]: modify orderButtonsClicked method logic

- orderButtonsClicked 안에 중복 showCurrentStock() 메서드 호출 제거
- viewTransitionButton -> moveToStockViewButton 으로 네이밍 변경 및 아울렛 재설정

stevenkim18 commented:
코멘트 단 내용 이외에 크게 수정할 부분들은 없어 보입니다!

배열로 관리했던 뷰들을 하나씩 나누신 것은 잘하신 것 같아요!

어제 구두로 말했을 때 델리게이트 패턴을 사용하신다고 하셨는데 그떄 싱글톤 패턴해서 한개의 로컬 인스턴스 만으로 관리 되게 구현해보시면 더 좋을 것 같습니다!! 너무 고생많으셨어요!!

머지하겠습니다!!!!
